### PR TITLE
InputCommon/OSX: Refactor IOKit controller interface

### PR DESCRIPTION
**Pull request by spycrab**
Cleaner, (probably) quicker and safer.

Fixes (on Xbox One Wireless, maybe does something for other controllers too): 
* Select button not being detected
* Triggers not being detected

### CODE REVIEW

File: Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.h [hunk 1]
```diff
@@ -83,6 +85,9 @@ class Joystick : public ForceFeedback::ForceFeedbackDevice
 private:
   const IOHIDDeviceRef m_device;
   const std::string m_device_name;
+  std::vector<IOHIDElementCookie> m_cookies;
```
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.h]: This doesn't make sense as a member variable as it's only needed during construction. You could declare it in the ctor and pass it to `AddElements` as `known_cookies`. Also your usage would maybe be more suited by a `std::set` as you only check for duplicates. The long `std::find_if(...)` call would then just be `known_cookies.count(cookie)`.

File: Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm [hunk 2]
```diff
@@ -2,75 +2,71 @@
...
-  if (axes)
-  {
-    std::vector<IOHIDElementRef> elems;
-    for (int i = 0; i < CFArrayGetCount(axes); i++)
-    {
-      IOHIDElementRef e = (IOHIDElementRef)CFArrayGetValueAtIndex(axes, i);
-      // DeviceElementDebugPrint(e, nullptr);
-      uint32_t usage = IOHIDElementGetUsage(e);
+    if (it != m_cookies.end())
+      break;
```
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm]: Shouldn't this `break` be a `continue` so the rest of the elements are looked at?

File: Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm [hunk 3]
```diff
@@ -2,75 +2,71 @@
...
-    @kIOHIDElementTypeKey : @(kIOHIDElementTypeInput_Button),
-    @kIOHIDElementUsagePageKey : @(kHIDPage_Button)
-  };
+  for (int i = 0; i < CFArrayGetCount(elements); i++)
+  {
+    IOHIDElementRef e = (IOHIDElementRef)CFArrayGetValueAtIndex(elements, i);
 
-  CFArrayRef buttons =
-      IOHIDDeviceCopyMatchingElements(m_device, (CFDictionaryRef)buttonDict, kIOHIDOptionsTypeNone);
+    uint32_t usage = IOHIDElementGetUsage(e);
```
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm]: Looks like these could be `const` and you could declare `usage`/`usage_page` down lower where they are actually used.

File: Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm [hunk 4]
```diff
@@ -2,75 +2,71 @@
...
+void Joystick::AddElements(CFArrayRef elements)
 {
-  // Buttons
-  NSDictionary* buttonDict = @{
-    @kIOHIDElementTypeKey : @(kIOHIDElementTypeInput_Button),
-    @kIOHIDElementUsagePageKey : @(kHIDPage_Button)
-  };
+  for (int i = 0; i < CFArrayGetCount(elements); i++)
+  {
+    IOHIDElementRef e = (IOHIDElementRef)CFArrayGetValueAtIndex(elements, i);
```
Comment (jordan-woyak) [inline Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm]: Ugly c-cast. (I know it was there before)

Comment (spycrab) [reply Source/Core/InputCommon/ControllerInterface/OSX/OSXJoystick.mm]: C-style casts are fine in Objective C FWIW.

**Pull request merged**